plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/simulate/ProcessedData.java

```diff
@@ -0,0 +1,141 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.plugin.ingest.simulate;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Streamable;
+import org.elasticsearch.common.xcontent.StatusToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentBuilderString;
+import org.elasticsearch.ingest.Data;
+import org.elasticsearch.rest.RestStatus;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+public class ProcessedData implements Streamable, StatusToXContent {
+
+    private String processorId;
+    private Data data;
+    private Throwable failure;
+
+    public ProcessedData() {
+
+    }
+
+    public ProcessedData(String processorId, Data data) {
+        this.processorId = processorId;
+        this.data = data;
+    }
+
+    public ProcessedData(Throwable failure) {
+        this.failure = failure;
+    }
+
+    public boolean isFailed() {
+        return this.failure != null;
+    }
+
+    public Data getData() {
+        return data;
+    }
+
+    public String getProcessorId() {
+        return processorId;
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        boolean isFailure = in.readBoolean();
+        if (isFailure) {
+            this.failure = in.readThrowable();
+            // TODO(talevy): check out mget for throwable limitations
+        } else {
+            this.processorId = in.readString();
+            String index = in.readString();
+            String type = in.readString();
+            String id = in.readString();
+            Map<String, Object> doc = in.readMap();
+            this.data = new Data(index, type, id, doc);
+        }
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeBoolean(isFailed());
+        if (isFailed()) {
+            out.writeThrowable(failure);
+        } else {
+            out.writeString(processorId);
+            out.writeString(data.getIndex());
+            out.writeString(data.getType());
+            out.writeString(data.getId());
+            out.writeMap(data.getDocument());
+        }
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(Fields.PROCESSOR_ID, processorId);
+        builder.field(Fields.ERROR, isFailed());
+        if (isFailed()) {
+            builder.field(Fields.FAILURE, failure.toString());
+        } else {
+            builder.field(Fields.MODIFIED, data.isModified());
+            builder.field(Fields.DOCUMENT, data.getDocument());
```

Comment:
oh I just realized that I suggested below to make Data implement both Streamable and ToXContent...which is not correct as it would mess up dependencies. I think sooner or later we should extract these classes to a different module to prevent mistakes and confusion and enforce the no deps rule that we follow. We should have some shared code around outputting Data to ToXContent and serializing it two. Shall we have an enhanced version of Data as part of the plugin that accepts a Data instance but also implements ToXContent and Streamable?